xarray/core/coordinates.py

```diff
@@ -796,6 +797,126 @@ def _ipython_key_completions_(self):
+    # and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    # This will add both the coordinates variables and their index.
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
+
+    _data: DataTree
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        variables = self._data._data_variables | self._data._coord_variables
+        return Frozen({k: v for k, v in variables.items() if k in self._names})
+
+    def __getitem__(self, key: Hashable) -> DataArray:
+        if key in self._data._data_variables or key in self._data.children:
+            raise KeyError(key)
+        return self._data[key]
+
+    def to_dataset(self) -> Dataset:
+        """Convert these coordinates into a new Dataset"""
+        return self._data.to_dataset()._copy_listed(self._names)
+
+    def _update_coords(
+        self, coords: dict[Hashable, Variable], indexes: Mapping[Any, Index]
+    ) -> None:
+
+        # TODO I don't know how to update coordinates that live in parent nodes
+        # TODO We would have to find the correct node and update `._node_coord_variables`
+
+        coord_variables = self._data._coord_variables.copy()
+        coord_variables.update(coords)
+
+        # check for inconsistent state *before* modifying anything in-place
+        variables = coord_variables | self._data._data_variables.copy()
+        # TODO is there a subtlety here with rebuild_dims?
+        dims = calculate_dimensions(variables)
+        new_coord_names = set(coords)
+        for dim, size in dims.items():
+            if dim in variables:
+                new_coord_names.add(dim)
+
+        # TODO we need to upgrade these variables to coord variables somehow
+        # coord_variables.update(new_coord_names)
+
+        self._data._coord_variables = coord_variables
+        self._data._dims = dims
```

Comment:
All these assignment should be to the `._node....` variables (you can't assign to `._coord_variables`, etc)